php/class-customize-snapshot-manager.php

```diff
@@ -0,0 +1,559 @@
+	protected $contextual_post_data;
+
+	/**
+	 * Customize_Snapshot instance.
+	 *
+	 * @access protected
+	 * @var Customize_Snapshot
+	 */
+	protected $snapshot;
+
+	/**
+	 * Constructor.
+	 *
+	 * @access public
+	 *
+	 * @param Plugin $plugin
+	 */
+	public function __construct( Plugin $plugin ) {
+		// Bail if our conditions are not met.
+		if ( ! ( ( isset( $_REQUEST['wp_customize'] ) && 'on' == $_REQUEST['wp_customize'] )
+			|| ( is_admin() && 'customize.php' == basename( $_SERVER['PHP_SELF'] ) )
+			|| ( isset( $_REQUEST['customize_snapshot_uuid'] ) )
+		) ) {
+			return;
+		}
+
+		$this->plugin = $plugin;
+
+		if ( ! did_action( 'setup_theme' ) ) {
+			// Note that Customize_Snapshot::populate_customized_post_var() happens next at priority 1.
+			add_action( 'setup_theme', array( $this, 'store_post_data' ), 0 );
+		} else {
+			$this->store_post_data();
+		}
+
+		$uuid = isset( $_REQUEST['customize_snapshot_uuid'] ) ? $_REQUEST['customize_snapshot_uuid'] : null;
+		$scope = isset( $_REQUEST['scope'] ) ? $_REQUEST['scope'] : 'dirty';
+		$apply_dirty = ( 'dirty' === $scope );
+
+		// Bootstrap the Customizer.
+		if ( empty( $GLOBALS['wp_customize'] ) || ! ( $GLOBALS['wp_customize'] instanceof \WP_Customize_Manager ) && $uuid ) {
+			require_once( ABSPATH . WPINC . '/class-wp-customize-manager.php' );
+			$GLOBALS['wp_customize'] = new \WP_Customize_Manager();
+		}
+
+		$this->snapshot = new Customize_Snapshot( $GLOBALS['wp_customize'], $uuid, $apply_dirty );
+
+		add_action( 'init', array( $this, 'maybe_force_redirect' ), 0 );
+		add_action( 'init', array( $this, 'create_post_type' ), 0 );
+		add_action( 'customize_controls_enqueue_scripts', array( $this, 'enqueue_scripts' ) );
+		add_action( 'wp_ajax_' . self::AJAX_ACTION, array( $this, 'update_snapshot' ) );
+		add_action( 'customize_save_after', array( $this, 'save_snapshot' ) );
+		add_action( 'admin_bar_menu', array( $this, 'customize_menu' ), 41 );
+		add_action( 'customize_controls_print_footer_scripts', array( $this, 'render_templates' ) );
+
+		// Preview a Snapshot
+		add_action( 'after_setup_theme', array( $this, 'set_post_values' ), 1 );
+		add_action( 'wp_loaded', array( $this, 'preview' ) );
+	}
+
+	/**
+	 * Get the current URL.
+	 *
+	 * @return string
+	 */
+	public function current_url() {
+		return ( is_ssl() ? 'https://' : 'http://' ) . $_SERVER['HTTP_HOST'] . $_SERVER['REQUEST_URI'];
+	}
+
+	/**
+	 * Redirect when preview is not allowed for the current theme.
+	 */
+	public function maybe_force_redirect() {
+		if ( false === $this->snapshot->is_preview() && isset( $_GET['customize_snapshot_uuid'] ) ) {
+			$args = array( 'customize_snapshot_uuid', 'scope' );
+			$current_url = esc_url( remove_query_arg( $args, $this->current_url() ) );
+			wp_safe_redirect( $current_url );
+			exit;
+		}
+	}
+
+	/**
+	 * Decode and store any $_POST['snapshot_customized'] data.
+	 *
+	 * The value is used by Customize_Snapshot_Manager::save().
+	 */
+	public function store_post_data() {
```

Comment:
I don't see where this method does a redirect. If the user is not authenticated, then we should not be reading anything from `$_POST`. If the user is not authenticated, we should only ever accept data read from the snapshot post, as the UUID keys into values that have been previously stored by an authenticated user.